README.md
# Root Detector

A library to check devices using root access or simply root/jailbreak checking.

These are the current checks/tricks we are using to give an indication of root.

And there are some parameters when checking root:

| Params      | Description |
| ----------- | ----------- |
| busyBox      | By default `busyBox` is **false**, And when the value is true then it checks root with busyBox which is usually used on some android devices |
| ignoreSimulator   | By default `ignoreSimulator` is **false** , And when the value is true then it checks root with busyBox which is usually used on some android devices |


## Setup

- Android

No config needed on Android

- iOS

Add this code in the Info.plist file whose position is in the /ios/Runner/ folder

```xml
<key>LSApplicationQueriesSchemes</key>
	<array>
	<string>cydia</string>
</array>
```

## Usage

The first step you need to install this dependency into pubspec.yaml in your Flutter project.

```yaml
dependencies:
  root_detector: // recommended to use lastest version
```

### checkIsRoot

```dart
try {
 final result = await RootDetector.isRooted(
   busyBox: true, // by default is false
   ignoreSimulator: true, // by default is false
 ); // type data is bool
 return result;
} on PlatformException catch(e){
  // TODO: handling your error, whenever have error from native code
}
```

More example in [github](https://github.com/wisnuwiry/root_detector/tree/main/example).

## Credit

- [RootBeer](https://github.com/scottyab/rootbeer)

CHANGELOG.md
## 0.0.1

* 🎉 initial release.

## 0.0.2

* Fix channel name
* Change `checkIsRootedWithBusyBox` to `isRootedWithBusyBox`

## 0.0.3

* Migrate to null safety

### 0.0.4

* Implement Cek JailBreak in IOS

### 0.0.5

* Update Rootbear version

### 0.0.6

* Implement in iOS
* Add feature ignoreSimulator

test/root_detector_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('root_detector');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  // test('get', () async {
  //   expect(await RootDetector.platformVersion, '42');
  // });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';
import 'package:root_detector_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());
  });
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:root_detector/root_detector.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _isRooted = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      await RootDetector.isRooted(
        busyBox: true,
        ignoreSimulator: false,
      ).then((value) {
        setState(() {
          _isRooted = value.toString();
        });
      });
    } on PlatformException {
      setState(() {
        _isRooted = 'Failed to get root status.';
      });
    }
  }

  @override
  void setState(VoidCallback fn) {
    if (mounted) {
      super.setState(fn);
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Root Detector'),
        ),
        body: Center(
          child: Text('Device is Rooted: $_isRooted\n'),
        ),
      ),
    );
  }
}


ios/Classes/JailBreakHelper.swift
//
//  JailBreakHelper.swift
//  root_detector
//
//  Created by Wisnu Saputra on 01/12/21.
//

import Foundation

private struct JailBreakHelper {
    //check if cydia is installed (using URI Scheme)
    static func hasCydiaInstalled() -> Bool {
        return UIApplication.shared.canOpenURL(URL(string: "cydia://")!)
    }
    
    //Check if suspicious apps (Cydia, FakeCarrier, Icy etc.) is installed
    static func isContainsSuspiciousApps() -> Bool {
        for path in suspiciousAppsPathToCheck {
            if FileManager.default.fileExists(atPath: path) {
                return true
            }
        }
        return false
    }
    
    //Check if system contains suspicious files
    static func isSuspiciousSystemPathsExists() -> Bool {
        for path in suspiciousSystemPathsToCheck {
            if FileManager.default.fileExists(atPath: path) {
                return true
            }
        }
        return false
    }
    
    //Check if app can edit system files
    static func canEditSystemFiles() -> Bool {
        let jailBreakText = "Developer Insider"
        do {
            try jailBreakText.write(toFile: jailBreakText, atomically: true, encoding: .utf8)
            return true
        } catch {
            return false
        }
    }
    
    //suspicious apps path to check
    static var suspiciousAppsPathToCheck: [String] {
        return ["/Applications/Cydia.app",
                "/Applications/blackra1n.app",
                "/Applications/FakeCarrier.app",
                "/Applications/Icy.app",
                "/Applications/IntelliScreen.app",
                "/Applications/MxTube.app",
                "/Applications/RockApp.app",
                "/Applications/SBSettings.app",
                "/Applications/WinterBoard.app"
        ]
    }
    
    //suspicious system paths to check
    static var suspiciousSystemPathsToCheck: [String] {
        return ["/Library/MobileSubstrate/DynamicLibraries/LiveClock.plist",
                "/Library/MobileSubstrate/DynamicLibraries/Veency.plist",
                "/private/var/lib/apt",
                "/private/var/lib/apt/",
                "/private/var/lib/cydia",
                "/private/var/mobile/Library/SBSettings/Themes",
                "/private/var/stash",
                "/private/var/tmp/cydia.log",
                "/System/Library/LaunchDaemons/com.ikey.bbot.plist",
                "/System/Library/LaunchDaemons/com.saurik.Cydia.Startup.plist",
                "/usr/bin/sshd",
                "/usr/libexec/sftp-server",
                "/usr/sbin/sshd",
                "/etc/apt",
                "/bin/bash",
                "/Library/MobileSubstrate/MobileSubstrate.dylib"
        ]
    }
}

extension UIDevice {
    var isSimulator: Bool {
        return TARGET_OS_SIMULATOR != 0
    }
    
    var isJailBroken: Bool {
        get {
            if UIDevice.current.isSimulator { return false }
            if JailBreakHelper.hasCydiaInstalled() { return true }
            if JailBreakHelper.isContainsSuspiciousApps() { return true }
            if JailBreakHelper.isSuspiciousSystemPathsExists() { return true }
            return JailBreakHelper.canEditSystemFiles()
        }
    }
}


ios/Classes/SwiftRootDetectorPlugin.swift
import Flutter
import UIKit

public class SwiftRootDetectorPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "space.wisnuwiry/root_detector", binaryMessenger: registrar.messenger())
        let instance = SwiftRootDetectorPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if(call.method == "checkIsRooted"){
            let isJailBroken = UIDevice.current.isJailBroken
            let isSimulator = UIDevice.current.isSimulator
            let args = call.arguments
            
            if let ignoreSimulator = args as? Bool {
                if ignoreSimulator && isSimulator && isJailBroken {
                    result(false)
                }else{
                    result(isJailBroken)
                }
            }else{
                result(isJailBroken)
            }
        }
    }
}


android/src/main/kotlin/space/wisnuwiry/root_detector/RootDetectorPlugin.kt
package space.wisnuwiry.root_detector

import android.content.Context
import androidx.annotation.NonNull
import com.scottyab.rootbeer.RootBeer

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.lang.IllegalArgumentException
import android.os.Build

/** RootDetectorPlugin */
class RootDetectorPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var rootBeer: RootBeer

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel =
            MethodChannel(flutterPluginBinding.binaryMessenger, "space.wisnuwiry/root_detector")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        rootBeer = RootBeer(context)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val ignoreSimulator = call.arguments == true

        if (call.method == "checkIsRooted") {
            checkIsRooted(result, ignoreSimulator)
        } else if (call.method == "checkIsRootedWithBusyBox") {
            checkIsRootedWithBusyBox(result, ignoreSimulator)
        } else {
            result.notImplemented()
        }
    }


    // Get check root status in normal devices
    private fun checkIsRooted(@NonNull result: Result, ignoreSimulator: Boolean) {
        try {
            val isRoot = rootBeer.isRooted
            if (isRoot && ignoreSimulator && isEmulator()) {
                result.success(false)
            } else {
                result.success(isRoot)
            }
        } catch (e: IllegalArgumentException) {
            result.error(e.message, e.message, e.stackTrace)
        }
    }

    // Get check root status in devices have the busybox binary
    // Example devices have a busybox:
    // - All OnePlus Devices
    // - Moto E
    // - OPPO R9m (ColorOS 3.0,Android 5.1,Android security patch January 5, 2018 )
    private fun checkIsRootedWithBusyBox(@NonNull result: Result, ignoreSimulator: Boolean) {
        try {
            val isRoot = rootBeer.isRootedWithBusyBoxCheck
            if (isRoot && ignoreSimulator && isEmulator()) {
                result.success(false)
            } else {
                result.success(isRoot)
            }
        } catch (e: IllegalArgumentException) {
            result.error(e.message, e.message, e.stackTrace)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {}

    override fun onDetachedFromActivityForConfigChanges() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}

    override fun onDetachedFromActivity() {}

    /**
     * A simple emulator-detection based on the flutter tools detection logic and a couple of legacy
     * detection systems
     */
    private fun isEmulator(): Boolean {
        return (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic")
                || Build.FINGERPRINT.startsWith("generic")
                || Build.FINGERPRINT.startsWith("unknown")
                || Build.HARDWARE.contains("goldfish")
                || Build.HARDWARE.contains("ranchu")
                || Build.MODEL.contains("google_sdk")
                || Build.MODEL.contains("Emulator")
                || Build.MODEL.contains("Android SDK built for x86")
                || Build.MANUFACTURER.contains("Genymotion")
                || Build.PRODUCT.contains("sdk_google")
                || Build.PRODUCT.contains("google_sdk")
                || Build.PRODUCT.contains("sdk")
                || Build.PRODUCT.contains("sdk_x86")
                || Build.PRODUCT.contains("vbox86p")
                || Build.PRODUCT.contains("emulator")
                || Build.PRODUCT.contains("simulator"))
    }
}


lib/root_detector.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';

class RootDetector {
  static const MethodChannel _channel =
      const MethodChannel("space.wisnuwiry/root_detector");

  /// To check status device is rooted or not
  ///
  /// For now this work only in Android, this method can return:
  ///
  /// - `true` => when device is rooted
  /// - `false` => when device is not rooted
  ///
  /// You can do error handling if an error occurs from native code,
  /// and you can catch error code with [PlatformException]
  ///
  /// The following devices are known the have the [busyBox] binary present on the stock rom:
  ///
  /// - All OnePlus Devices
  /// - Moto E
  /// - OPPO R9m (ColorOS 3.0,Android 5.1, Android security patch January 5, 2018 )
  ///
  /// you can ignore the root check / jailbreak when the device is not
  /// a real device (Simulator / Emulator) by `ignoreSimulator` make it `true`
  ///
  /// `ignoreSimulator = true`
  ///
  /// Example for usage:
  ///
  /// ```dart
  /// try {
  ///  final result = await RootDetector.isRooted(); // type data is bool
  ///  return result;
  /// } on PlatformException catch(e){
  ///  // TODO: handling your error, whenever have error from native code
  /// }
  /// ```
  static Future<bool> isRooted({
    /// Params [busyBox] works only for Android
    bool busyBox = false,
    bool ignoreSimulator = false,
  }) async {
    if (busyBox == true && Platform.isAndroid) {
      return await _channel.invokeMethod(
        'checkIsRootedWithBusyBox',
        ignoreSimulator,
      );
    }

    final result = await _channel.invokeMethod(
      'checkIsRooted',
      ignoreSimulator,
    );
    return result;
  }

  /// Use `isRooted(busyBox: true)` method instead of `isRootedWithBusyBox`
  /// and to check jailbreak with busyBox
  ///
  /// Example:
  ///
  /// /// ```dart
  /// try {
  ///  final result = await RootDetector.isRooted(busyBox: true); // type data is bool
  ///  return result;
  /// } on PlatformException catch(e){
  ///  // TODO: handling your error, whenever have error from native code
  /// }
  /// ```
  ///
  @deprecated
  static Future<bool> get isRootedWithBusyBox async {
    final bool result = await _channel.invokeMethod('checkIsRootedWithBusyBox');
    return result;
  }
}
